folx/experimental/pallas/forward_laplacian.py

```diff
@@ -0,0 +1,437 @@
+    # TODO: Can we avoid calling `.dense_array` on the Jacobians and instead use sparse matrices here?
+    # This would help with LapNet-like attention
+    x, jacobian, laplacian = kernel_fn(
+        q.x,
+        q.jacobian.dense_array,
+        q.laplacian,
+        k.x,
+        k.jacobian.dense_array,
+        k.laplacian,
+        v.x,
+        v.jacobian.dense_array,
+        v.laplacian,
+        mask,
+        input_mask,
+    )
+    return FwdLaplArray(x, FwdJacobian(jacobian, None), laplacian)
+
+
+@jax.jit
+def folx_mha_forward_laplacian_kernel(
+    q: jax.Array,
+    q_jac: jax.Array,
+    q_lap: jax.Array,
+    k: jax.Array,
+    k_jac: jax.Array,
+    k_lap: jax.Array,
+    v: jax.Array,
+    v_jac: jax.Array,
+    v_lap: jax.Array,
+    mask: jax.Array,
+    input_mask: jax.Array,
+) -> Tuple[jax.Array, jax.Array, jax.Array]:
+    r"""Vanilla folx implementation of the multi-head attention forward Laplacian kernel."""
+    q_fwd_lap = FwdLaplArray(
+        q, FwdJacobian.from_dense(q_jac * input_mask[..., None, None, None]), q_lap
+    )
+    k_fwd_lap = FwdLaplArray(
+        k, FwdJacobian.from_dense(k_jac * input_mask[..., None, None, None]), k_lap
+    )
+    v_fwd_lap = FwdLaplArray(
+        v, FwdJacobian.from_dense(v_jac * input_mask[..., None, None, None]), v_lap
+    )
+    o_fwd_lap = forward_laplacian(reference_mha_kernel)(q_fwd_lap, k_fwd_lap, v_fwd_lap, mask)  # type: ignore
+    return o_fwd_lap.x, o_fwd_lap.jacobian.dense_array, o_fwd_lap.laplacian
+
+
+@jax.jit
+def reference_mha_forward_laplacian_kernel(
+    q: jax.Array,
+    q_jac: jax.Array,
+    q_lap: jax.Array,
+    k: jax.Array,
+    k_jac: jax.Array,
+    k_lap: jax.Array,
+    v: jax.Array,
+    v_jac: jax.Array,
+    v_lap: jax.Array,
+    mask: jax.Array,
+    input_mask: jax.Array,
+):
+    r"""Reference jax implementation of the multi-head attention forward Laplacian kernel."""
+    # [batch_size, seq_len, num_heads, seq_len]
+    square_mask = mask[:, None, None, :] * mask[:, :, None, None]
+    # [input_dim, batch_size, seq_len, num_heads, head_dim]
+    coordinate_and_electron_mask = input_mask[:, :, None, None, None] * mask[None, :, :, None, None]
+    # [batch_size, seq_len, num_heads, head_dim]
+    qkv_mask = mask[:, :, None, None]
+    q = jnp.where(qkv_mask, q, 0.0)
+    k = jnp.where(qkv_mask, k, 0.0)
+    v = jnp.where(qkv_mask, v, 0.0)
+    q_jac = jnp.where(coordinate_and_electron_mask, q_jac, 0.0)
+    k_jac = jnp.where(coordinate_and_electron_mask, k_jac, 0.0)
+    v_jac = jnp.where(coordinate_and_electron_mask, v_jac, 0.0)
+    q_lap = jnp.where(qkv_mask, q_lap, 0.0)
+    k_lap = jnp.where(qkv_mask, k_lap, 0.0)
+    v_lap = jnp.where(qkv_mask, v_lap, 0.0)
+    # Forward
+    s = jnp.einsum("Bnhd,BNhd->BnhN", q, k)
+    s = jnp.where(square_mask, s, -1e20)
+    p = jax.nn.softmax(s, axis=-1)
+    o = jnp.einsum("BnhN,BNhd->Bnhd", p, v)
+
+    # Jacobian
```

Comment:
While I am a huge fan of einsum, but here we potentially leave performance on the table. The reason is that we are at the mercy of opt_einsum yielding the same order every time. I can't tell whether that's an issue here but here's an example:
Imagine I want to compute
ABC and DAB
If I select the orders
(AB)C and (DA)B, I have to do 4 matmuls in total.
If I were to order them: 
(AB)C and D(AB), I only need 3 mamuls, since I can reuse AB.

In standard JAX, if I write down the first version, I get 4 matmuls. The latter (without explicit caching of AB) would yield 3 matmuls due to common suppression elimination.

So here we implicitly hope that opt_einsum yields us with the minimal number of total matmuls (which it is not guaranteeing or considering in any way).Note that in some cases doing more matmuls may still be faster to avoid materializing large low-rank tensors (which is what opt_einsum optimizes).

Reply:
An example:
```py
import jax
import jax.numpy as jnp

A, B, C, D = jax.random.normal(jax.random.PRNGKey(0), (4, 10, 10))


def f1(A, B, C, D):
    return A @ B @ C, D @ A @ B


print("f1")
lowered = jax.jit(f1).lower(A, B, C, D)
compiled = lowered.compile()
print(compiled.as_text())
print(compiled.cost_analysis()[0]["flops"])


def f2(A, B, C, D):
    return A @ B @ C, D @ (A @ B)


print("f2")
lowered = jax.jit(f2).lower(A, B, C, D)
compiled = lowered.compile()
print(compiled.as_text())
print(compiled.cost_analysis()[0]["flops"])
```

```
f1
HloModule jit_f1, is_scheduled=true, entry_computation_layout={(f32[10,10]{1,0}, f32[10,10]{1,0}, f32[10,10]{1,0}, f32[10,10]{1,0})->(f32[10,10]{1,0}, f32[10,10]{1,0})}, allow_spmd_sharding_propagation_to_parameters={true,true,true,true}, allow_spmd_sharding_propagation_to_output={true,true}

ENTRY %main.10 (Arg_0.1: f32[10,10], Arg_1.2: f32[10,10], Arg_2.3: f32[10,10], Arg_3.4: f32[10,10]) -> (f32[10,10], f32[10,10]) {
  %Arg_0.1 = f32[10,10]{1,0} parameter(0), metadata={op_name="A"}
  %Arg_1.2 = f32[10,10]{1,0} parameter(1), metadata={op_name="B"}
  %dot.5 = f32[10,10]{1,0} dot(f32[10,10]{1,0} %Arg_0.1, f32[10,10]{1,0} %Arg_1.2), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_name="jit(f1)/jit(main)/dot_general" source_file="/Users/gaoni/Documents/Repositories/test/hello.py" source_line=8}
  %Arg_2.3 = f32[10,10]{1,0} parameter(2), metadata={op_name="C"}
  %dot.6 = f32[10,10]{1,0} dot(f32[10,10]{1,0} %dot.5, f32[10,10]{1,0} %Arg_2.3), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_name="jit(f1)/jit(main)/dot_general" source_file="/Users/gaoni/Documents/Repositories/test/hello.py" source_line=8}
  %Arg_3.4 = f32[10,10]{1,0} parameter(3), metadata={op_name="D"}
  %dot.7 = f32[10,10]{1,0} dot(f32[10,10]{1,0} %Arg_3.4, f32[10,10]{1,0} %Arg_0.1), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_name="jit(f1)/jit(main)/dot_general" source_file="/Users/gaoni/Documents/Repositories/test/hello.py" source_line=8}
  %dot.8 = f32[10,10]{1,0} dot(f32[10,10]{1,0} %dot.7, f32[10,10]{1,0} %Arg_1.2), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_name="jit(f1)/jit(main)/dot_general" source_file="/Users/gaoni/Documents/Repositories/test/hello.py" source_line=8}
  ROOT %tuple.9 = (f32[10,10]{1,0}, f32[10,10]{1,0}) tuple(f32[10,10]{1,0} %dot.6, f32[10,10]{1,0} %dot.8)
}


8000.0
f2
HloModule jit_f2, is_scheduled=true, entry_computation_layout={(f32[10,10]{1,0}, f32[10,10]{1,0}, f32[10,10]{1,0}, f32[10,10]{1,0})->(f32[10,10]{1,0}, f32[10,10]{1,0})}, allow_spmd_sharding_propagation_to_parameters={true,true,true,true}, allow_spmd_sharding_propagation_to_output={true,true}

ENTRY %main.10 (Arg_0.1: f32[10,10], Arg_1.2: f32[10,10], Arg_2.3: f32[10,10], Arg_3.4: f32[10,10]) -> (f32[10,10], f32[10,10]) {
  %Arg_0.1 = f32[10,10]{1,0} parameter(0), metadata={op_name="A"}
  %Arg_1.2 = f32[10,10]{1,0} parameter(1), metadata={op_name="B"}
  %dot.5 = f32[10,10]{1,0} dot(f32[10,10]{1,0} %Arg_0.1, f32[10,10]{1,0} %Arg_1.2), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_name="jit(f2)/jit(main)/dot_general" source_file="/Users/gaoni/Documents/Repositories/test/hello.py" source_line=19}
  %Arg_2.3 = f32[10,10]{1,0} parameter(2), metadata={op_name="C"}
  %dot.6 = f32[10,10]{1,0} dot(f32[10,10]{1,0} %dot.5, f32[10,10]{1,0} %Arg_2.3), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_name="jit(f2)/jit(main)/dot_general" source_file="/Users/gaoni/Documents/Repositories/test/hello.py" source_line=19}
  %Arg_3.4 = f32[10,10]{1,0} parameter(3), metadata={op_name="D"}
  %dot.8 = f32[10,10]{1,0} dot(f32[10,10]{1,0} %Arg_3.4, f32[10,10]{1,0} %dot.5), lhs_contracting_dims={1}, rhs_contracting_dims={0}, metadata={op_name="jit(f2)/jit(main)/dot_general" source_file="/Users/gaoni/Documents/Repositories/test/hello.py" source_line=19}
  ROOT %tuple.9 = (f32[10,10]{1,0}, f32[10,10]{1,0}) tuple(f32[10,10]{1,0} %dot.6, f32[10,10]{1,0} %dot.8)
}


6000.0
```

Reply:
I haven't benchmarked the reference implementation

Reply:
I'm generally fine with both solutions but I agree that a user should not accidentally use it. :)

After resolving this one, I'd merge the PR.

Reply:
Ok, then I added some warning messages [here](https://github.com/microsoft/folx/pull/26/commits/32a5c07509b326d324822f231852421353178cc3).